### ISSUE TITLE
Mixin causes an re-entrance error after loading Mixin plugins

### ISSUE THREAD

**Issue by Minecrell** [SpongePowered/Mixin#139]
I've run into a quite weird problem which seems to only happen in some very special cases. Basically, it seems like Mixin is causing a re-entrance error itself **after** loading the Mixin config plugin classes.

Since it is a bit hard to explain, I've created a small test project which fails with a re-entrance error when running it: https://github.com/Minecrell/mixintest

This is the error I get: https://gist.github.com/Minecrell/723c1289b96c5e9e09581ae018005ba8

A few notes about the project and the problem:

- The project can be imported as Gradle project and then run in the IDE with `net.minecraft.launchwrapper.Launch` as main class and the program arguments `--tweakClass net.minecrell.test.launch.TestTweaker`

- There is a [`MainClass`] which calls a method in a [`SecondClass`]. Both classes have a Mixin targeting the class which must be applied successfully for the project to run without an error.

- A mixin config plugin must be present for the error to occur.
- If I remove the call to [`SecondClass.doNothing()` in `MixinMainClass`](https://github.com/Minecrell/mixintest/blob/master/src/main/java/net/minecrell/test/mixin/core/MixinMainClass.java#L13), or remove the Mixin for [`SecondClass`] it works fine. So it will only happen if Mixin has to apply another Mixin **after** the first Mixin was applied.
- I tried to debug to find the cause of the problem. What I noticed is that it seems like Mixin fails to load [`SecondClass`] because the re-entrance state is not reset after loading the Mixin config classes:

  - Mixin begins transforming [`MainClass`]
  - Mixin config plugins are being loaded (that will cause re-entrance since Mixin will load them inside the `transform` method) --> `ReEntranceState.semaphore` is set to true.
  - Mixin config plugin class has not a Mixin so the re-entrance is silently ignored
  - Mixins for [`MainClass`] are successfully applied
  - Mixin begins transforming [`SecondClass`] (at this point the `ReEntranceState.semaphore` is still set to true from the mixin config class, and since [`SecondClass`] has a Mixin it will fail with a re-entrance error)

So tl;dr: I think the `ReEntranceState` is not properly reset after re-entrance has been ignored (when no Mixin was targetting the class). I could imagine it works in Sponge because the `ReEntranceState.clear()` method is called before the second class is transformed.

[`MainClass`]: https://github.com/Minecrell/mixintest/blob/master/src/main/java/net/minecrell/test/MainClass.java
[`SecondClass`]: https://github.com/Minecrell/mixintest/blob/master/src/main/java/net/minecrell/test/SecondClass.java

**Comment by Mumfrey** [comment]
I think I understand what you're getting at, I'll investigate and see what I can do.

**Comment by Mumfrey** [comment]
So `semaphore` is needed to check whether any re-entrance happened since `clear()` was called, and thus cannot simply by unset once the depth is properly popped back to zero. However, using it for the actual depth-check seems to be the issue. Basically I should be using `check()` in the transformer and not `isSet()`. This should rectify the problem.

**Comment by Minecrell** [comment]
Thanks, that fixes the issue. :)
